drivers/base/bootvid/i386/pc98/bootvid.c

```diff
@@ -0,0 +1,464 @@
+                SetPixel(X, Y, (UCHAR)TextColor);
+            else if (BackColor != BV_COLOR_NONE)
+                SetPixel(X, Y, (UCHAR)BackColor);
+        }
+    }
+}
+
+VOID
+NTAPI
+PreserveRow(
+    _In_ ULONG CurrentTop,
+    _In_ ULONG TopDelta,
+    _In_ BOOLEAN Restore)
+{
+    PUCHAR OldPosition, NewPosition;
+    ULONG PixelCount = TopDelta * SCREEN_WIDTH;
+
+    if (Restore)
+    {
+        /* Restore the row by copying back the contents saved off-screen */
+        OldPosition = (PUCHAR)(FrameBuffer + FB_OFFSET(0, SCREEN_HEIGHT));
+        NewPosition = (PUCHAR)(FrameBuffer + FB_OFFSET(0, CurrentTop));
+    }
+    else
+    {
+        /* Preserve the row by saving its contents off-screen */
+        OldPosition = (PUCHAR)(FrameBuffer + FB_OFFSET(0, CurrentTop));
+        NewPosition = (PUCHAR)(FrameBuffer + FB_OFFSET(0, SCREEN_HEIGHT));
+    }
+
+    while (PixelCount--)
+        WRITE_REGISTER_UCHAR(NewPosition++, READ_REGISTER_UCHAR(OldPosition++));
+}
+
+VOID
+PrepareForSetPixel(VOID)
+{
+    NOTHING;
+}
+
+VOID
+NTAPI
+DoScroll(
+    _In_ ULONG Scroll)
+{
+    USHORT i, Line;
+    PUCHAR Src, Dst;
+    PULONG SrcWide, DstWide;
+    USHORT PixelCount = VidpScrollRegion[2] - VidpScrollRegion[0];
+    ULONG_PTR SourceOffset = FrameBuffer + FB_OFFSET(VidpScrollRegion[0], VidpScrollRegion[1]);
+    ULONG_PTR DestinationOffset = FrameBuffer + FB_OFFSET(VidpScrollRegion[0], VidpScrollRegion[1] - Scroll);
+
+    for (Line = VidpScrollRegion[1]; Line <= VidpScrollRegion[3]; Line++)
+    {
+        SrcWide = (PULONG)SourceOffset;
+        DstWide = (PULONG)DestinationOffset;
+        for (i = 0; i <= PixelCount / sizeof(ULONG); i++)
+            WRITE_REGISTER_ULONG(DstWide++, READ_REGISTER_ULONG(SrcWide++));
+
+        Src = (PUCHAR)SrcWide;
+        Dst = (PUCHAR)DstWide;
+        for (i = 0; i <= PixelCount % sizeof(ULONG); i++)
+            WRITE_REGISTER_UCHAR(Dst++, READ_REGISTER_UCHAR(Src++));
+
+        SourceOffset += SCREEN_WIDTH;
+        DestinationOffset += SCREEN_WIDTH;
+    }
+}
+
+/* PUBLIC FUNCTIONS ***********************************************************/
+
+BOOLEAN
+NTAPI
+VidInitialize(
+    _In_ BOOLEAN SetMode)
+{
+    PHYSICAL_ADDRESS BaseAddress;
+
+    BaseAddress.QuadPart = VRAM_NORMAL_PLANE_I;
+    VideoMemoryI = (ULONG_PTR)MmMapIoSpace(BaseAddress, VRAM_PLANE_SIZE, MmNonCached);
+    if (!VideoMemoryI)
+        goto Failure;
+
+    if (!HasPegcController())
+        goto Failure;
+
+    BaseAddress.QuadPart = PEGC_FRAMEBUFFER_PACKED;
+    FrameBuffer = (ULONG_PTR)MmMapIoSpace(BaseAddress, PEGC_FRAMEBUFFER_SIZE, MmNonCached);
+    if (!FrameBuffer)
+        goto Failure;
+
+    if (SetMode)
+        VidResetDisplay(TRUE);
+
+    return TRUE;
+
+Failure:
+    if (!VideoMemoryI) MmUnmapIoSpace((PVOID)VideoMemoryI, VRAM_NORMAL_PLANE_I);
+    if (!FrameBuffer) MmUnmapIoSpace((PVOID)FrameBuffer, PEGC_FRAMEBUFFER_SIZE);
+
+    return FALSE;
+}
+
+VOID
+NTAPI
+VidCleanUp(VOID)
+{
+    NOTHING;
```

Comment:
No cleanup? (calling VidResetDisplay and unmapping memory)

Reply:
Not sure. The `VidInitialize` function is called exactly once. What happens if you press the `TAB+K` multiple times?

Reply:
Check what kind of things the regular VGA boot driver does.